plugins/zenoh-plugin-storage-manager/src/replica/mod.rs

```diff
@@ -0,0 +1,291 @@
+    replica_config: ReplicaConfig,
+    digests_published: RwLock<HashSet<u64>>, // checksum of all digests generated and published by this replica
+}
+
+impl Replica {
+    // This function starts the replica by initializing all the components
+    pub async fn start(
+        config: ReplicaConfig,
+        session: Arc<Session>,
+        storage: Box<dyn zenoh_backend_traits::Storage>,
+        in_interceptor: Option<Arc<dyn Fn(Sample) -> Sample + Send + Sync>>,
+        out_interceptor: Option<Arc<dyn Fn(Sample) -> Sample + Send + Sync>>,
+        key_expr: OwnedKeyExpr,
+        name: &str,
+    ) -> ZResult<Sender<crate::StorageMessage>> {
+        trace!("[REPLICA]Opening session...");
+        let startup_entries = storage.get_all_entries().await?;
+
+        let replica = Replica {
+            name: name.to_string(),
+            session,
+            key_expr,
+            replica_config: config,
+            digests_published: RwLock::new(HashSet::new()),
+        };
+
+        // Create channels for communication between components
+        // channel to queue digests to be aligned
+        let (tx_digest, rx_digest) = flume::unbounded();
+        // channel for aligner to send missing samples to storage
+        let (tx_sample, rx_sample) = flume::unbounded();
+        // channel for storage to send logging information back
+        let (tx_log, rx_log) = flume::unbounded();
+
+        let config = replica.replica_config.clone();
+        // snapshotter
+        let snapshotter =
+            Arc::new(Snapshotter::new(rx_log, startup_entries.clone(), config.clone()).await);
+        // digest sub
+        let digest_sub = replica.start_digest_sub(tx_digest);
+        // queryable for alignment
+        let digest_key = Replica::get_digest_key(replica.key_expr.clone(), config.align_prefix);
+        let align_queryable = AlignQueryable::start_align_queryable(
+            replica.session.clone(),
+            digest_key.clone(),
+            &replica.name,
+            snapshotter.clone(),
+        );
+        // aligner
+        let aligner = Aligner::start_aligner(
+            replica.session.clone(),
+            digest_key,
+            rx_digest,
+            tx_sample,
+            snapshotter.clone(),
+        );
+        // digest pub
+        let digest_pub = replica.start_digest_pub(snapshotter.clone());
+
+        //updating snapshot time
+        let snapshot_task = snapshotter.start();
+
+        //actual storage
+        let replication = ReplicationService {
+            empty_start: startup_entries.is_empty(),
+            aligner_updates: rx_sample,
+            log_propagation: tx_log,
+        };
+        let storage_task = StorageService::start(
+            replica.session.clone(),
+            replica.key_expr.clone(),
+            &replica.name,
+            storage,
+            in_interceptor,
+            out_interceptor,
+            Some(replication),
+        );
+
+        let result = join!(
```

Comment:
This `join` will block until all the tasks are done. But there is no mechanism to stop the tasks, except storage_task.
The storage_task must stop the other tasks when ending.
And this operation must not wait for the tasks to end.

Reply:
Using `select` instead of join. Hence when the `storage_task` is killed through a signal, it stops the entire replica as well.